// ng2-pdf-viewer/esm2015/src/app/pdf-viewer/pdf-viewer.module.js
/**
 * Created by vadimdez on 01/11/2016.
 */
import { NgModule } from '@angular/core';
import { PdfViewerComponent } from './pdf-viewer.component';
import * as ɵngcc0 from '@angular/core';
export * from './typings';
export class PdfViewerModule {
}
PdfViewerModule.ɵfac = function PdfViewerModule_Factory(t) { return new (t || PdfViewerModule)(); };
PdfViewerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: PdfViewerModule });
PdfViewerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PdfViewerModule, [{
        type: NgModule,
        args: [{
                declarations: [PdfViewerComponent],
                exports: [PdfViewerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(PdfViewerModule, { declarations: function () { return [PdfViewerComponent]; }, exports: function () { return [PdfViewerComponent]; } }); })();
