// @angular/fire/esm2015/storage/storage.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: storage.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { GetDownloadURLPipeModule } from './pipes/storageUrl.pipe';
import { AngularFireStorage } from './storage';
import * as ɵngcc0 from '@angular/core';
export class AngularFireStorageModule {
}
AngularFireStorageModule.ɵfac = function AngularFireStorageModule_Factory(t) { return new (t || AngularFireStorageModule)(); };
AngularFireStorageModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AngularFireStorageModule });
AngularFireStorageModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [AngularFireStorage], imports: [GetDownloadURLPipeModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AngularFireStorageModule, [{
        type: NgModule,
        args: [{
                exports: [GetDownloadURLPipeModule],
                providers: [AngularFireStorage]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AngularFireStorageModule, { exports: function () { return [GetDownloadURLPipeModule]; } }); })();
