// ngx-daterange/esm2015/components/manual-inputs/manual-inputs.component.js
import { Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import * as momentNs from 'moment';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from '../../pipes/format-moment-date.pipe';

function ManualInputsComponent_ng_container_0_div_1_div_2_i_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "i", 10);
    ɵngcc0.ɵɵtext(1, "event");
    ɵngcc0.ɵɵelementEnd();
} }
function ManualInputsComponent_ng_container_0_div_1_div_2_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 11);
} }
function ManualInputsComponent_ng_container_0_div_1_div_2_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1, "From:\u00A0");
    ɵngcc0.ɵɵelementEnd();
} }
const _c0 = function (a1) { return { "d-flex align-items-center input-with-icon": true, "material": a1 }; };
const _c1 = function () { return { standalone: true }; };
function ManualInputsComponent_ng_container_0_div_1_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 6);
    ɵngcc0.ɵɵtemplate(2, ManualInputsComponent_ng_container_0_div_1_div_2_i_2_Template, 2, 0, "i", 7);
    ɵngcc0.ɵɵtemplate(3, ManualInputsComponent_ng_container_0_div_1_div_2_i_3_Template, 1, 0, "i", 8);
    ɵngcc0.ɵɵtemplate(4, ManualInputsComponent_ng_container_0_div_1_div_2_span_4_Template, 2, 0, "span", 0);
    ɵngcc0.ɵɵelementStart(5, "input", 9);
    ɵngcc0.ɵɵlistener("blur", function ManualInputsComponent_ng_container_0_div_1_div_2_Template_input_blur_5_listener($event) { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(3); return ctx_r9.setDateFromInput($event, true); });
    ɵngcc0.ɵɵpipe(6, "formatMomentDate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(9, _c0, ctx_r2.options.icons === "material"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.options.icons === "material");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.options.icons === "font-awesome");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.fromDate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", ɵngcc0.ɵɵpipeBind2(6, 6, ctx_r2.fromDate, ctx_r2.options.format))("ngModelOptions", ɵngcc0.ɵɵpureFunction0(11, _c1));
} }
function ManualInputsComponent_ng_container_0_div_1_input_3_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "input", 9);
    ɵngcc0.ɵɵlistener("blur", function ManualInputsComponent_ng_container_0_div_1_input_3_Template_input_blur_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(3); return ctx_r11.setDateFromInput($event, true); });
    ɵngcc0.ɵɵpipe(1, "formatMomentDate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngModel", ɵngcc0.ɵɵpipeBind2(1, 2, ctx_r3.fromDate, ctx_r3.options.format))("ngModelOptions", ɵngcc0.ɵɵpureFunction0(5, _c1));
} }
function ManualInputsComponent_ng_container_0_div_1_div_5_i_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "i", 10);
    ɵngcc0.ɵɵtext(1, "event");
    ɵngcc0.ɵɵelementEnd();
} }
function ManualInputsComponent_ng_container_0_div_1_div_5_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 11);
} }
function ManualInputsComponent_ng_container_0_div_1_div_5_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1, "To:\u00A0");
    ɵngcc0.ɵɵelementEnd();
} }
function ManualInputsComponent_ng_container_0_div_1_div_5_Template(rf, ctx) { if (rf & 1) {
    const _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 6);
    ɵngcc0.ɵɵtemplate(2, ManualInputsComponent_ng_container_0_div_1_div_5_i_2_Template, 2, 0, "i", 7);
    ɵngcc0.ɵɵtemplate(3, ManualInputsComponent_ng_container_0_div_1_div_5_i_3_Template, 1, 0, "i", 8);
    ɵngcc0.ɵɵtemplate(4, ManualInputsComponent_ng_container_0_div_1_div_5_span_4_Template, 2, 0, "span", 0);
    ɵngcc0.ɵɵelementStart(5, "input", 12);
    ɵngcc0.ɵɵlistener("blur", function ManualInputsComponent_ng_container_0_div_1_div_5_Template_input_blur_5_listener($event) { ɵngcc0.ɵɵrestoreView(_r17); const ctx_r16 = ɵngcc0.ɵɵnextContext(3); return ctx_r16.setDateFromInput($event); });
    ɵngcc0.ɵɵpipe(6, "formatMomentDate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(9, _c0, ctx_r4.options.icons === "material"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.options.icons === "material");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.options.icons === "font-awesome");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.toDate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", ɵngcc0.ɵɵpipeBind2(6, 6, ctx_r4.toDate, ctx_r4.options.format))("ngModelOptions", ɵngcc0.ɵɵpureFunction0(11, _c1));
} }
function ManualInputsComponent_ng_container_0_div_1_input_6_Template(rf, ctx) { if (rf & 1) {
    const _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "input", 13);
    ɵngcc0.ɵɵlistener("blur", function ManualInputsComponent_ng_container_0_div_1_input_6_Template_input_blur_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r19); const ctx_r18 = ɵngcc0.ɵɵnextContext(3); return ctx_r18.setDateFromInput($event); });
    ɵngcc0.ɵɵpipe(1, "formatMomentDate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngModel", ɵngcc0.ɵɵpipeBind2(1, 2, ctx_r5.toDate, ctx_r5.options.format))("ngModelOptions", ɵngcc0.ɵɵpureFunction0(5, _c1));
} }
function ManualInputsComponent_ng_container_0_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementStart(1, "div", 3);
    ɵngcc0.ɵɵtemplate(2, ManualInputsComponent_ng_container_0_div_1_div_2_Template, 7, 12, "div", 0);
    ɵngcc0.ɵɵtemplate(3, ManualInputsComponent_ng_container_0_div_1_input_3_Template, 2, 6, "input", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 3);
    ɵngcc0.ɵɵtemplate(5, ManualInputsComponent_ng_container_0_div_1_div_5_Template, 7, 12, "div", 0);
    ɵngcc0.ɵɵtemplate(6, ManualInputsComponent_ng_container_0_div_1_input_6_Template, 2, 6, "input", 5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.options.icons !== "default");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.options.icons === "default");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.options.icons !== "default");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.options.icons === "default");
} }
function ManualInputsComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ManualInputsComponent_ng_container_0_div_1_Template, 7, 4, "div", 1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.options.singleCalendar);
} }
const moment = momentNs;
export class ManualInputsComponent {
    constructor() {
        this.fromDate = null;
        this.toDate = null;
        this.dateFromInput = new EventEmitter();
    }
    ngOnInit() {
    }
    setDateFromInput(event, isLeft = false) {
        this.dateFromInput.emit({ event, isLeft });
    }
}
ManualInputsComponent.ɵfac = function ManualInputsComponent_Factory(t) { return new (t || ManualInputsComponent)(); };
ManualInputsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ManualInputsComponent, selectors: [["manual-inputs"]], inputs: { fromDate: "fromDate", toDate: "toDate", options: "options", isMobile: "isMobile" }, outputs: { dateFromInput: "dateFromInput" }, decls: 1, vars: 1, consts: [[4, "ngIf"], ["class", "row form-inputs mb-0", 4, "ngIf"], [1, "row", "form-inputs", "mb-0"], [1, "col"], ["class", "form-control", "type", "text", "name", "daterangepicker_start", "placeholder", "Select From Date", 3, "ngModel", "ngModelOptions", "blur", 4, "ngIf"], ["class", "form-control", "name", "daterangepicker_end", 3, "ngModel", "ngModelOptions", "blur", 4, "ngIf"], [3, "ngClass"], ["class", "material-icons", 4, "ngIf"], ["class", "far fa-calendar-alt", 4, "ngIf"], ["type", "text", "name", "daterangepicker_start", "placeholder", "Select From Date", 1, "form-control", 3, "ngModel", "ngModelOptions", "blur"], [1, "material-icons"], [1, "far", "fa-calendar-alt"], ["type", "text", "name", "daterangepicker_end", "placeholder", "Select To Date", 1, "form-control", 3, "ngModel", "ngModelOptions", "blur"], ["name", "daterangepicker_end", 1, "form-control", 3, "ngModel", "ngModelOptions", "blur"]], template: function ManualInputsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ManualInputsComponent_ng_container_0_Template, 2, 1, "ng-container", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isMobile);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgClass, ɵngcc2.DefaultValueAccessor, ɵngcc2.NgControlStatus, ɵngcc2.NgModel], pipes: [ɵngcc3.FormatMomentDatePipe], styles: [".is-chrome[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{position:relative;top:1px}.form-inputs[_ngcontent-%COMP%]{background-color:#6e777c;border-top-left-radius:8px;border-top-right-radius:8px;color:#fff;font-size:16px}.form-inputs[_ngcontent-%COMP%] > .col[_ngcontent-%COMP%]{min-width:270px;padding:.5rem 15px}.form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]{align-content:flex-start;align-items:center!important;display:flex}.form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]   i[_ngcontent-%COMP%]{color:#fff;margin-right:.5rem}.form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]   i.material-icons[_ngcontent-%COMP%]{top:0}.form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{line-height:1}.form-inputs[_ngcontent-%COMP%]   .form-control[_ngcontent-%COMP%]{background:none;border:0;color:#fff;display:inline-block;padding:0;width:150px}.form-inputs[_ngcontent-%COMP%]   .form-control[_ngcontent-%COMP%]::-moz-placeholder{color:#fff}.form-inputs[_ngcontent-%COMP%]   .form-control[_ngcontent-%COMP%]::placeholder{color:#fff}[_ngcontent-%COMP%]:not(.mobile)   .form-inputs[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]:nth-child(2), [_ngcontent-%COMP%]:not(.mobile)   .row-calendars[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]:nth-child(2){border-left:2px solid #c6d1d6}"] });
ManualInputsComponent.ctorParameters = () => [];
ManualInputsComponent.propDecorators = {
    options: [{ type: Input }],
    fromDate: [{ type: Input }],
    toDate: [{ type: Input }],
    isMobile: [{ type: Input }],
    dateFromInput: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ManualInputsComponent, [{
        type: Component,
        args: [{
                encapsulation: ViewEncapsulation.Emulated,
                selector: 'manual-inputs',
                template: "<ng-container *ngIf=\"!isMobile\">\n\n\t<div class=\"row form-inputs mb-0\" *ngIf=\"!options.singleCalendar\">\n\t\t<div class=\"col\">\n\t\t\t<div *ngIf=\"options.icons !== 'default'\">\n\t\t\t\t<div [ngClass]=\"{ 'd-flex align-items-center input-with-icon': true, 'material': options.icons === 'material' }\">\n\t\t\t\t\t<i *ngIf=\"options.icons === 'material'\" class=\"material-icons\">event</i>\n\t\t\t\t\t<i *ngIf=\"options.icons === 'font-awesome'\" class=\"far fa-calendar-alt\"></i>\n\t\t\t\t\t<span *ngIf=\"fromDate\">From:&nbsp;</span>\n\t\t\t\t\t<input class=\"form-control\" [ngModel]=\"fromDate | formatMomentDate: options.format\" (blur)=\"setDateFromInput($event, true)\" type=\"text\" name=\"daterangepicker_start\" [ngModelOptions]=\"{ standalone: true }\" placeholder=\"Select From Date\" />\n\t\t\t\t</div>\n\t\t\t</div>\n\t\n\t\t\t<input *ngIf=\"options.icons === 'default'\" class=\"form-control\" [ngModel]=\"fromDate | formatMomentDate: options.format\" (blur)=\"setDateFromInput($event, true)\" type=\"text\" name=\"daterangepicker_start\" [ngModelOptions]=\"{standalone: true}\" placeholder=\"Select From Date\" />\n\t\t</div>\n\t\n\t\t<div class=\"col\">\n\t\t\t<div *ngIf=\"options.icons !== 'default'\">\n\t\t\t\t\t<div [ngClass]=\"{ 'd-flex align-items-center input-with-icon': true, 'material': options.icons === 'material' }\">\n\t\t\t\t\t\t<i *ngIf=\"options.icons === 'material'\" class=\"material-icons\">event</i>\n\t\t\t\t\t\t<i *ngIf=\"options.icons === 'font-awesome'\" class=\"far fa-calendar-alt\"></i>\n\t\t\t\t\t\t<span *ngIf=\"toDate\">To:&nbsp;</span>\n\t\t\t\t\t\t<input class=\"form-control\" [ngModel]=\"toDate | formatMomentDate: options.format\" (blur)=\"setDateFromInput($event)\" type=\"text\" name=\"daterangepicker_end\" [ngModelOptions]=\"{ standalone: true }\" placeholder=\"Select To Date\" />\n\t\t\t\t\t</div>\n\t\t\t\t</div>\n\t\n\t\t\t<input *ngIf=\"options.icons === 'default'\" class=\"form-control\" [ngModel]=\"toDate | formatMomentDate: options.format\" (blur)=\"setDateFromInput($event)\" name=\"daterangepicker_end\" [ngModelOptions]=\"{standalone: true}\" />\n\t\t</div>\n\t</div>\n</ng-container>",
                styles: [".is-chrome .form-inputs .input-with-icon span{position:relative;top:1px}.form-inputs{background-color:#6e777c;border-top-left-radius:8px;border-top-right-radius:8px;color:#fff;font-size:16px}.form-inputs>.col{min-width:270px;padding:.5rem 15px}.form-inputs .input-with-icon{align-content:flex-start;align-items:center!important;display:flex}.form-inputs .input-with-icon i{color:#fff;margin-right:.5rem}.form-inputs .input-with-icon i.material-icons{top:0}.form-inputs .input-with-icon span{line-height:1}.form-inputs .form-control{background:none;border:0;color:#fff;display:inline-block;padding:0;width:150px}.form-inputs .form-control::-moz-placeholder{color:#fff}.form-inputs .form-control::placeholder{color:#fff}:not(.mobile) .form-inputs>div:nth-child(2),:not(.mobile) .row-calendars>div:nth-child(2){border-left:2px solid #c6d1d6}"]
            }]
    }], function () { return []; }, { fromDate: [{
            type: Input
        }], toDate: [{
            type: Input
        }], dateFromInput: [{
            type: Output
        }], options: [{
            type: Input
        }], isMobile: [{
            type: Input
        }] }); })();
