// ngx-daterange/esm2015/components/datepicker/date-range-picker.component.js
import { Component, Input, Output, EventEmitter, HostListener, ViewEncapsulation } from '@angular/core';
import { FormGroup, FormControl } from '@angular/forms';
import { defaultDateFormat, defaultDateRangePickerOptions } from '../../constants';
import * as momentNs from 'moment';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from '../manual-inputs/manual-inputs.component';
import * as ɵngcc4 from '../calendar/calendar.component';
import * as ɵngcc5 from '../predefined-date-ranges/predefined-date-ranges.component';

function DateRangePickerComponent_div_0_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "label");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(3, "input", 23);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("for", ctx_r2.instanceId);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r2.options.labelText);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("formControlName", ctx_r2.controlName);
    ɵngcc0.ɵɵattribute("id", ctx_r2.instanceId);
} }
const _c0 = function (a1, a2) { return { "col-12": true, "col-sm-6": a1, "col__mobile": a2 }; };
function DateRangePickerComponent_div_0_div_18_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵelementStart(1, "calendar", 24);
    ɵngcc0.ɵɵlistener("dateChanged", function DateRangePickerComponent_div_0_div_18_Template_calendar_dateChanged_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(2); return ctx_r7.dateChanged($event); })("monthChanged", function DateRangePickerComponent_div_0_div_18_Template_calendar_monthChanged_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r9 = ɵngcc0.ɵɵnextContext(2); return ctx_r9.monthChanged($event); })("yearChanged", function DateRangePickerComponent_div_0_div_18_Template_calendar_yearChanged_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r10 = ɵngcc0.ɵɵnextContext(2); return ctx_r10.yearChanged($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(9, _c0, !ctx_r3.isMobile, ctx_r3.isMobile));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("month", ctx_r3.toMonth)("year", ctx_r3.toYear)("format", ctx_r3.options.format)("selectedFromDate", ctx_r3.fromDate)("selectedToDate", ctx_r3.toDate)("minDate", ctx_r3.options.minDate)("maxDate", ctx_r3.options.maxDate)("icons", ctx_r3.options.icons);
} }
function DateRangePickerComponent_div_0_button_23_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 25);
    ɵngcc0.ɵɵlistener("click", function DateRangePickerComponent_div_0_button_23_Template_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(2); return ctx_r11.reset($event); });
    ɵngcc0.ɵɵtext(1, "Reset");
    ɵngcc0.ɵɵelementEnd();
} }
function DateRangePickerComponent_div_0_button_24_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 26);
    ɵngcc0.ɵɵlistener("click", function DateRangePickerComponent_div_0_button_24_Template_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r14); const ctx_r13 = ɵngcc0.ɵɵnextContext(2); return ctx_r13.apply($event); });
    ɵngcc0.ɵɵtext(1, "Apply");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("disabled", !ctx_r5.enableApplyButton);
} }
function DateRangePickerComponent_div_0_div_33_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵelementStart(1, "calendar", 24);
    ɵngcc0.ɵɵlistener("dateChanged", function DateRangePickerComponent_div_0_div_33_Template_calendar_dateChanged_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r15 = ɵngcc0.ɵɵnextContext(2); return ctx_r15.dateChanged($event); })("monthChanged", function DateRangePickerComponent_div_0_div_33_Template_calendar_monthChanged_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r17 = ɵngcc0.ɵɵnextContext(2); return ctx_r17.monthChanged($event); })("yearChanged", function DateRangePickerComponent_div_0_div_33_Template_calendar_yearChanged_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r18 = ɵngcc0.ɵɵnextContext(2); return ctx_r18.yearChanged($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(9, _c0, !ctx_r6.isMobile, ctx_r6.isMobile));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("month", ctx_r6.toMonth)("year", ctx_r6.toYear)("format", ctx_r6.options.format)("selectedFromDate", ctx_r6.fromDate)("selectedToDate", ctx_r6.toDate)("minDate", ctx_r6.options.minDate)("maxDate", ctx_r6.options.maxDate)("icons", ctx_r6.options.icons);
} }
const _c1 = function (a0) { return { "modal-display": a0 }; };
const _c2 = function (a1, a2, a3) { return { "dateRangePickerModal": true, "dateRangePicker-single": a1, "is-chrome": a2, "mobile": a3 }; };
const _c3 = function (a1) { return { "col-12": true, "col-sm-6": a1 }; };
const _c4 = function (a1) { return { "calendar": true, "calendar-single": a1 }; };
const _c5 = function (a1, a2, a3, a4, a5, a6) { return { "dateRangePicker": true, "invisible": a1, "dateRangePicker-single": a2, "is-chrome": a3, "open-left": a4, "open-right": a5, "mobile": a6 }; };
function DateRangePickerComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵelementStart(1, "div", null, 2);
    ɵngcc0.ɵɵprojection(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 3);
    ɵngcc0.ɵɵtemplate(5, DateRangePickerComponent_div_0_ng_container_5_Template, 4, 4, "ng-container", 4);
    ɵngcc0.ɵɵelementStart(6, "div", 5);
    ɵngcc0.ɵɵelementStart(7, "div", 6);
    ɵngcc0.ɵɵelementStart(8, "div", 7);
    ɵngcc0.ɵɵelementStart(9, "div", 8);
    ɵngcc0.ɵɵelementStart(10, "div", 9);
    ɵngcc0.ɵɵelementStart(11, "div", 10);
    ɵngcc0.ɵɵelementStart(12, "div", 11);
    ɵngcc0.ɵɵelementStart(13, "div", 12);
    ɵngcc0.ɵɵelementStart(14, "manual-inputs", 13);
    ɵngcc0.ɵɵlistener("dateFromInput", function DateRangePickerComponent_div_0_Template_manual_inputs_dateFromInput_14_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r19 = ɵngcc0.ɵɵnextContext(); return ctx_r19.setDateFromInput($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(15, "div", 14);
    ɵngcc0.ɵɵelementStart(16, "div", 9);
    ɵngcc0.ɵɵelementStart(17, "calendar", 15);
    ɵngcc0.ɵɵlistener("monthChanged", function DateRangePickerComponent_div_0_Template_calendar_monthChanged_17_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r21 = ɵngcc0.ɵɵnextContext(); return ctx_r21.monthChanged($event); })("yearChanged", function DateRangePickerComponent_div_0_Template_calendar_yearChanged_17_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r22 = ɵngcc0.ɵɵnextContext(); return ctx_r22.yearChanged($event); })("dateChanged", function DateRangePickerComponent_div_0_Template_calendar_dateChanged_17_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r23 = ɵngcc0.ɵɵnextContext(); return ctx_r23.dateChanged($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(18, DateRangePickerComponent_div_0_div_18_Template, 2, 12, "div", 16);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(19, "predefined-date-ranges", 17);
    ɵngcc0.ɵɵlistener("closeEvent", function DateRangePickerComponent_div_0_Template_predefined_date_ranges_closeEvent_19_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r24 = ɵngcc0.ɵɵnextContext(); return ctx_r24.close($event); })("resetEvent", function DateRangePickerComponent_div_0_Template_predefined_date_ranges_resetEvent_19_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r25 = ɵngcc0.ɵɵnextContext(); return ctx_r25.reset($event); })("applyEvent", function DateRangePickerComponent_div_0_Template_predefined_date_ranges_applyEvent_19_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r26 = ɵngcc0.ɵɵnextContext(); return ctx_r26.apply($event); })("predefinedRange", function DateRangePickerComponent_div_0_Template_predefined_date_ranges_predefinedRange_19_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r27 = ɵngcc0.ɵɵnextContext(); return ctx_r27.applyPredefinedRange($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(20, "div", 18);
    ɵngcc0.ɵɵelementStart(21, "button", 19);
    ɵngcc0.ɵɵlistener("click", function DateRangePickerComponent_div_0_Template_button_click_21_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r28 = ɵngcc0.ɵɵnextContext(); return ctx_r28.close($event); });
    ɵngcc0.ɵɵtext(22, "Close");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(23, DateRangePickerComponent_div_0_button_23_Template, 2, 0, "button", 20);
    ɵngcc0.ɵɵtemplate(24, DateRangePickerComponent_div_0_button_24_Template, 2, 1, "button", 21);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(25, "div", 9);
    ɵngcc0.ɵɵelementStart(26, "div", 10);
    ɵngcc0.ɵɵelementStart(27, "div", 11);
    ɵngcc0.ɵɵelementStart(28, "div", 12);
    ɵngcc0.ɵɵelementStart(29, "manual-inputs", 13);
    ɵngcc0.ɵɵlistener("dateFromInput", function DateRangePickerComponent_div_0_Template_manual_inputs_dateFromInput_29_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r29 = ɵngcc0.ɵɵnextContext(); return ctx_r29.setDateFromInput($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(30, "div", 14);
    ɵngcc0.ɵɵelementStart(31, "div", 9);
    ɵngcc0.ɵɵelementStart(32, "calendar", 15);
    ɵngcc0.ɵɵlistener("monthChanged", function DateRangePickerComponent_div_0_Template_calendar_monthChanged_32_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r30 = ɵngcc0.ɵɵnextContext(); return ctx_r30.monthChanged($event); })("yearChanged", function DateRangePickerComponent_div_0_Template_calendar_yearChanged_32_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r31 = ɵngcc0.ɵɵnextContext(); return ctx_r31.yearChanged($event); })("dateChanged", function DateRangePickerComponent_div_0_Template_calendar_dateChanged_32_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r32 = ɵngcc0.ɵɵnextContext(); return ctx_r32.dateChanged($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(33, DateRangePickerComponent_div_0_div_33_Template, 2, 12, "div", 16);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(34, "predefined-date-ranges", 22);
    ɵngcc0.ɵɵlistener("closeEvent", function DateRangePickerComponent_div_0_Template_predefined_date_ranges_closeEvent_34_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r33 = ɵngcc0.ɵɵnextContext(); return ctx_r33.close($event); })("resetEvent", function DateRangePickerComponent_div_0_Template_predefined_date_ranges_resetEvent_34_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r34 = ɵngcc0.ɵɵnextContext(); return ctx_r34.reset($event); })("applyEvent", function DateRangePickerComponent_div_0_Template_predefined_date_ranges_applyEvent_34_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r35 = ɵngcc0.ɵɵnextContext(); return ctx_r35.apply($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r1 = ɵngcc0.ɵɵreference(2);
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("formGroup", ctx_r0.parentFormGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", _r1.childNodes.length === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(47, _c1, ctx_r0.displayStyle));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction3(49, _c2, ctx_r0.options.singleCalendar, ctx_r0.checkChrome(), ctx_r0.isMobile));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("options", ctx_r0.options)("fromDate", ctx_r0.fromDate)("toDate", ctx_r0.toDate)("isMobile", ctx_r0.isMobile);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(53, _c3, !ctx_r0.isMobile && !ctx_r0.options.singleCalendar));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(55, _c4, ctx_r0.options.singleCalendar))("isLeft", true)("month", ctx_r0.fromMonth)("year", ctx_r0.fromYear)("format", ctx_r0.options.format)("selectedFromDate", ctx_r0.fromDate)("selectedToDate", ctx_r0.toDate)("minDate", ctx_r0.options.minDate)("maxDate", ctx_r0.options.maxDate)("singleCalendar", ctx_r0.options.singleCalendar)("icons", ctx_r0.options.icons);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.options.singleCalendar);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("options", ctx_r0.options)("range", ctx_r0.range)("enableApplyButton", ctx_r0.enableApplyButton);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.range && ctx_r0.options.showResetButton);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.options.autoApply);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction6(57, _c5, !ctx_r0.showCalendars, ctx_r0.options.singleCalendar, ctx_r0.checkChrome(), ctx_r0.options.position === "left", ctx_r0.options.position === "right", ctx_r0.isMobile));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("options", ctx_r0.options)("fromDate", ctx_r0.fromDate)("toDate", ctx_r0.toDate)("isMobile", ctx_r0.isMobile);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(64, _c3, !ctx_r0.isMobile && !ctx_r0.options.singleCalendar));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(66, _c4, ctx_r0.options.singleCalendar))("isLeft", true)("month", ctx_r0.fromMonth)("year", ctx_r0.fromYear)("format", ctx_r0.options.format)("selectedFromDate", ctx_r0.fromDate)("selectedToDate", ctx_r0.toDate)("minDate", ctx_r0.options.minDate)("maxDate", ctx_r0.options.maxDate)("singleCalendar", ctx_r0.options.singleCalendar)("icons", ctx_r0.options.icons);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.options.singleCalendar);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("options", ctx_r0.options)("range", ctx_r0.range)("enableApplyButton", ctx_r0.enableApplyButton);
} }
const _c6 = ["*"];
const moment = momentNs;
let instanceCount = 0;
export class DateRangePickerComponent {
    constructor() {
        this.options = defaultDateRangePickerOptions;
        this.controlName = 'dateRange';
        this.instanceId = null;
        this.fromDate = null;
        this.toDate = null;
        this.datepickerReset = new EventEmitter();
        this.rangeSelected = new EventEmitter();
        this.isMobile = false;
        this.range = '';
        this.showCalendars = false;
        this.displayStyle = false;
        if (!this.instanceId) {
            // assign auto-id
            this.instanceId = `dateRangePicker-${instanceCount++}`;
        }
    }
    get enableApplyButton() {
        let enabled = !this.options.autoApply && this.fromDate !== null;
        if (this.options.singleCalendar) {
            return enabled;
        }
        return enabled && this.toDate !== null;
    }
    handleClick(event) {
        var _a;
        const target = event.target;
        // close the DatePicker if clicking outside is not allowed
        if (!this.options.clickOutsideAllowed) {
            let targetPathClassNames = [];
            try {
                targetPathClassNames = event.composedPath().map((obj) => obj['className']) || [''];
            }
            catch (error) {
                // IE / Edge
                targetPathClassNames = (_a = event['path']) === null || _a === void 0 ? void 0 : _a.map(obj => obj.className);
            }
            let containerElementClassRoot = this.options.modal === true ? 'dateRangePickerModal' : 'dateRangePicker';
            const targetExistsInPath = targetPathClassNames === null || targetPathClassNames === void 0 ? void 0 : targetPathClassNames.some(className => {
                if (typeof className === 'string') {
                    return className && className.includes(containerElementClassRoot);
                }
                return false;
            });
            if (!targetExistsInPath) {
                this.toggleCalendarVisibility(false);
            }
        }
        // Close the DatePicker if the target input was clicked
        if (target.id === this.instanceId) {
            this.toggleCalendarVisibility(!this.showCalendars);
        }
    }
    ngOnInit() {
        if (navigator) {
            if (navigator.userAgent.match(/Android/i)
                || navigator.userAgent.match(/webOS/i)
                || navigator.userAgent.match(/iPhone/i)
                || navigator.userAgent.match(/iPod/i)
                || navigator.userAgent.match(/BlackBerry/i)
                || navigator.userAgent.match(/Windows Phone/i)) {
                this.isMobile = true;
            }
        }
        // ensure dates in options are valid
        this.validateOptionDates();
        // ensure input dates are within the min/max dates in options
        this.validateInputDates();
        if (this.options.preDefinedRanges && this.options.preDefinedRanges.length > 0) {
            this.defaultRanges = this.validateAndAssignPredefinedRanges(this.options.preDefinedRanges);
        }
        // assign values not present in options with default values
        const optionsKeys = Object.keys(this.options);
        const defaultValuesKeys = Object.keys(defaultDateRangePickerOptions);
        defaultValuesKeys.forEach((key) => {
            if (!optionsKeys.includes(key)) {
                this.options[key] = defaultDateRangePickerOptions[key];
            }
        });
        // update calendar grid
        this.updateCalendar();
        // create parent form group if it doesn't exist
        if (!this.parentFormGroup) {
            this.parentFormGroup = new FormGroup({});
        }
        // add form control to parent form group
        const value = this.formatRangeAsString();
        const control = new FormControl({ value, disabled: this.options.disableInputDisplay }, this.options.validators);
        if (this.options.disabled) {
            control.disable();
        }
        this.parentFormGroup.addControl(this.controlName, control);
        // set value of control
        this.setRange();
    }
    validateInputDates() {
        if (typeof this.fromDate === 'string') {
            this.fromDate = moment(this.fromDate);
        }
        if (typeof this.toDate === 'string') {
            this.toDate = moment(this.toDate);
        }
        if (this.fromDate && this.options.minDate && this.fromDate.isBefore(this.options.minDate, 'date')) {
            throw new RangeError('@Input fromDate is before the specified minDate in options');
        }
        if (this.toDate && this.options.maxDate && this.toDate.isAfter(this.options.maxDate, 'date')) {
            throw new RangeError('@Input toDate is after the specified maxDate in options');
        }
    }
    validateOptionDates() {
        // validate maxDate isn't before minDate or vice versa
        if (this.options) {
            if (this.options.minDate && this.options.maxDate) {
                if (this.options.minDate.isAfter(this.options.maxDate, 'date')) {
                    throw new RangeError('minDate specified in options is after the maxDate');
                }
                else if (this.options.maxDate.isBefore(this.options.minDate, 'date')) {
                    throw new RangeError('maxDate specified in options is before the minDate');
                }
            }
        }
    }
    // assists CSS to fix small positioning bug with From:/To: date text
    checkChrome() {
        return window['chrome'] ? 'is-chrome' : '';
    }
    toggleCalendarVisibility(value) {
        if (this.options.modal === true) {
            this.displayStyle = value !== null ? value === false ? false : true : this.showCalendars === false ? false : true;
        }
        else {
            this.showCalendars = value !== null ? value : !this.showCalendars;
        }
    }
    setFromToMonthYear(fromDate, toDate) {
        let tempFromDate = fromDate || this.fromDate || moment();
        let tempToDate = toDate || this.toDate || moment();
        if (!(tempFromDate instanceof moment)) {
            tempFromDate = moment(tempFromDate, defaultDateFormat);
        }
        if (!(tempToDate instanceof moment)) {
            tempToDate = moment(tempToDate, defaultDateFormat);
        }
        this.fromMonth = tempFromDate.get('month');
        this.fromYear = tempFromDate.get('year');
        this.toMonth = tempToDate.get('month');
        this.toYear = tempToDate.get('year');
    }
    updateCalendar() {
        // get month and year to show calendar
        this.setFromToMonthYear();
        this.setRange();
    }
    // update from/to based on selection
    dateChanged(changedData) {
        const value = changedData.day;
        const isLeft = changedData.isLeft;
        if (isLeft) {
            this.fromDate = value;
            if (this.fromDate.isAfter(this.toDate, 'date')) {
                this.toDate = this.fromDate.clone();
            }
        }
        else {
            this.toDate = value;
            if (this.toDate.isBefore(this.fromDate, 'date')) {
                this.fromDate = this.toDate.clone();
            }
        }
        this.setFromToMonthYear(this.fromDate, this.toDate);
        if (this.options.autoApply && (this.toDate || this.options.singleCalendar)) {
            this.toggleCalendarVisibility(false);
            this.setRange();
            this.emitRangeSelected();
        }
    }
    emitRangeSelected(data) {
        if (!data) {
            data = this.options.singleCalendar ? { start: this.fromDate } : { start: this.fromDate, end: this.toDate };
        }
        this.rangeSelected.emit(data);
    }
    getMoment(value) {
        return moment(value, this.options.format);
    }
    formatRangeAsString() {
        let range = '';
        if (this.options.singleCalendar && this.fromDate) {
            if (typeof this.fromDate === 'string') {
                this.fromDate = moment(this.fromDate);
            }
            range = this.fromDate.format(this.options.format);
        }
        else if (!this.options.singleCalendar && this.fromDate && this.toDate) {
            if (typeof this.fromDate === 'string') {
                this.fromDate = moment(this.fromDate);
            }
            if (typeof this.toDate === 'string') {
                this.toDate = moment(this.toDate);
            }
            range = `${this.fromDate.format(this.options.format)} - ${this.toDate.format(this.options.format)}`;
        }
        return range;
    }
    setRange() {
        var _a;
        this.range = this.formatRangeAsString();
        if (this.parentFormGroup) {
            const control = (_a = this.parentFormGroup) === null || _a === void 0 ? void 0 : _a.get(this.controlName);
            if (control) {
                control.setValue(this.range);
                control.updateValueAndValidity();
            }
        }
    }
    setDateFromInput(event) {
        const isLeft = event.isLeft;
        const target = event.event.target;
        try {
            if (target.value) {
                const day = this.getMoment(target.value);
                if (!day.isBefore(this.options.minDate) && !day.isAfter(this.options.maxDate)) {
                    if (isLeft && !this.fromDate) {
                        this.fromDate = day;
                    }
                    if (!isLeft && !this.toDate) {
                        this.toDate = day;
                    }
                    this.dateChanged({
                        day,
                        isLeft,
                    });
                    if (this.fromDate && this.toDate) {
                        this.setFromToMonthYear(this.fromDate, this.toDate);
                        if (!this.options.autoApply) {
                            this.emitRangeSelected();
                        }
                    }
                }
                else {
                    // assume nothing - reset values
                    this.fromDate = null;
                    this.toDate = null;
                    target.value = '';
                    target.focus();
                }
            }
        }
        catch (e) {
            console.error(e);
        }
    }
    monthChanged(data) {
        let temp;
        if (data.isLeft) {
            temp = moment([this.fromYear, this.fromMonth]).add(data.value, 'month');
            this.fromMonth = temp === null || temp === void 0 ? void 0 : temp.get('month');
            this.fromYear = temp === null || temp === void 0 ? void 0 : temp.get('year');
        }
        else {
            temp = moment([this.toYear, this.toMonth]).add(data.value, 'month');
            this.toMonth = temp === null || temp === void 0 ? void 0 : temp.get('month');
            this.toYear = temp === null || temp === void 0 ? void 0 : temp.get('year');
        }
    }
    yearChanged(data) {
        let temp;
        if (data.isLeft) {
            temp = moment([this.fromYear, this.fromMonth]).add(data.value, 'year');
            this.fromMonth = temp === null || temp === void 0 ? void 0 : temp.get('month');
            this.fromYear = temp === null || temp === void 0 ? void 0 : temp.get('year');
        }
        else {
            temp = moment([this.toYear, this.toMonth]).add(data.value, 'year');
            this.toMonth = temp === null || temp === void 0 ? void 0 : temp.get('month');
            this.toYear = temp === null || temp === void 0 ? void 0 : temp.get('year');
        }
    }
    close(event) {
        this.toggleCalendarVisibility(false);
        event.stopPropagation();
    }
    reset(event) {
        this.fromDate = null;
        this.toDate = null;
        this.setRange();
        this.datepickerReset.emit();
        event.stopPropagation();
    }
    apply(event) {
        this.toggleCalendarVisibility(false);
        this.setRange();
        this.emitRangeSelected();
        event.stopPropagation();
    }
    applyPredefinedRange(event) {
        // adjust to/from month/year so calendar months and years match range
        this.setFromToMonthYear(event.definedDateRange.value.start, event.definedDateRange.value.end);
        this.fromDate = event.definedDateRange.value.start;
        this.toDate = event.definedDateRange.value.end;
        if (this.options.autoApply) {
            this.apply(event.event);
        }
    }
    validateAndAssignPredefinedRanges(ranges) {
        return ranges.filter(range => {
            if (range.value.start.isAfter(range.value.end, 'date')) {
                throw new RangeError(`Pre-defined range "${range.name}" start date cannot be after the end date for the range.`);
            }
            if (this.options.minDate && range.value.start.isBefore(this.options.minDate)) {
                throw new RangeError(`Pre-defined range "${range.name}" start date is before the specified minDate in your options.`);
            }
            if (this.options.maxDate && range.value.end.isAfter(this.options.maxDate)) {
                throw new RangeError(`Pre-defined range "${range.name}" end date is after the specified maxDate in your options.`);
            }
            // add range to ranges
            return true;
        });
    }
}
DateRangePickerComponent.ɵfac = function DateRangePickerComponent_Factory(t) { return new (t || DateRangePickerComponent)(); };
DateRangePickerComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DateRangePickerComponent, selectors: [["date-range-picker"]], hostBindings: function DateRangePickerComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function DateRangePickerComponent_click_HostBindingHandler($event) { return ctx.handleClick($event); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, inputs: { options: "options", controlName: "controlName", instanceId: "instanceId", fromDate: "fromDate", toDate: "toDate", parentFormGroup: "parentFormGroup" }, outputs: { datepickerReset: "datepickerReset", rangeSelected: "rangeSelected" }, ngContentSelectors: _c6, decls: 1, vars: 1, consts: [["class", "dateRangePicker-wrapper", 4, "ngIf"], [1, "dateRangePicker-wrapper"], ["content", ""], [1, "form-group", 3, "formGroup"], [4, "ngIf"], ["tabindex", "-1", "role", "dialog", 1, "modal", 3, "ngClass"], ["role", "document", 1, "modal-dialog", "modal-dialog-centered"], [1, "modal-content"], [1, "modal-body"], [3, "ngClass"], [1, "container-fluid"], [1, "row"], [1, "col"], [3, "options", "fromDate", "toDate", "isMobile", "dateFromInput"], [1, "row", "row-calendars"], [3, "ngClass", "isLeft", "month", "year", "format", "selectedFromDate", "selectedToDate", "minDate", "maxDate", "singleCalendar", "icons", "monthChanged", "yearChanged", "dateChanged"], [3, "ngClass", 4, "ngIf"], [3, "options", "range", "enableApplyButton", "closeEvent", "resetEvent", "applyEvent", "predefinedRange"], [1, "modal-footer"], ["type", "button", 1, "btn", "btn-outline-secondary", "btn-sm", 3, "click"], ["type", "button", "class", "btn btn-outline-secondary btn-sm ml-3", 3, "click", 4, "ngIf"], ["type", "button", "class", "btn btn-primary btn-sm ml-3", 3, "disabled", "click", 4, "ngIf"], [3, "options", "range", "enableApplyButton", "closeEvent", "resetEvent", "applyEvent"], ["type", "text", 1, "form-control", 3, "formControlName"], [1, "calendar", 3, "month", "year", "format", "selectedFromDate", "selectedToDate", "minDate", "maxDate", "icons", "dateChanged", "monthChanged", "yearChanged"], ["type", "button", 1, "btn", "btn-outline-secondary", "btn-sm", "ml-3", 3, "click"], ["type", "button", 1, "btn", "btn-primary", "btn-sm", "ml-3", 3, "disabled", "click"]], template: function DateRangePickerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, DateRangePickerComponent_div_0_Template, 35, 68, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.parentFormGroup);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.NgControlStatusGroup, ɵngcc2.FormGroupDirective, ɵngcc1.NgClass, ɵngcc3.ManualInputsComponent, ɵngcc4.CalendarComponent, ɵngcc5.PredefinedDateRangesComponent, ɵngcc2.DefaultValueAccessor, ɵngcc2.NgControlStatus, ɵngcc2.FormControlName], styles: [".dateRangePicker-wrapper[_ngcontent-%COMP%]{border:none;position:relative}.dateRangePicker-wrapper[_ngcontent-%COMP%]   input.form-control[disabled][_ngcontent-%COMP%]{background-color:#fff}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .modal-body[_ngcontent-%COMP%]{padding:0!important}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .modal[_ngcontent-%COMP%]{-webkit-backdrop-filter:brightness(50%);backdrop-filter:brightness(50%)}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .modal-display[_ngcontent-%COMP%]{display:flex}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .modal-display[_ngcontent-%COMP%]   .modal-dialog[_ngcontent-%COMP%]{box-shadow:none;display:flex}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .modal-content[_ngcontent-%COMP%]{border:0;border-radius:8px;box-shadow:0 0 10px 0 rgba(129,139,144,.5);display:flex;width:602px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .modal-content[_ngcontent-%COMP%]   .modal-footer[_ngcontent-%COMP%]   .btn[_ngcontent-%COMP%]{width:auto}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]{border-top-left-radius:5px;border-top-right-radius:5px;font-family:Roboto,Helvetica,Arial,sans-serif;overflow:visible;z-index:3000}@media screen and (min-width:576px){.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]{min-width:600px}}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal.dateRangePicker-single[_ngcontent-%COMP%], .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal.mobile[_ngcontent-%COMP%]{min-width:300px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal.dateRangePicker-single[_ngcontent-%COMP%]:after, .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal.dateRangePicker-single[_ngcontent-%COMP%]:before, .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal.mobile[_ngcontent-%COMP%]:after, .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal.mobile[_ngcontent-%COMP%]:before{border-bottom-color:#fff}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal.is-chrome[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{position:relative;top:1px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]   .btn[_ngcontent-%COMP%], .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]   .btn.btn-link[_ngcontent-%COMP%]{min-width:60px;width:60px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]{background-color:#6e777c;border-top-left-radius:5px;border-top-right-radius:5px;color:#fff;font-size:16px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%] > .col[_ngcontent-%COMP%]{min-width:270px;padding:.5rem 15px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]{align-content:flex-start;align-items:center!important;display:flex}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]   i[_ngcontent-%COMP%]{color:#fff;margin-right:.5rem}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]   i.material-icons[_ngcontent-%COMP%]{top:0}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{line-height:1}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .form-control[_ngcontent-%COMP%]{background:none;border:0;color:#fff;display:inline-block;padding:0;width:150px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .form-control[_ngcontent-%COMP%]::-moz-placeholder{color:#fff}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .form-control[_ngcontent-%COMP%]::placeholder{color:#fff}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]   .row-1s[_ngcontent-%COMP%]   .col__mobile[_ngcontent-%COMP%]{border-top:1px dotted #ccc;margin-top:1rem}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]:not(.mobile)   .form-inputs[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]:nth-child(2), .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePickerModal[_ngcontent-%COMP%]:not(.mobile)   .row-calendars[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]:nth-child(2){border-left:2px solid #c6d1d6}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]{background:#fff;border-radius:9px;box-shadow:0 0 10px 0 rgba(129,139,144,.5);font-family:Roboto,Helvetica,Arial,sans-serif;overflow:visible;position:absolute;top:88px;z-index:3000}@media screen and (min-width:576px){.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]{min-width:600px}}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.dateRangePicker-single[_ngcontent-%COMP%], .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.mobile[_ngcontent-%COMP%]{min-width:300px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.dateRangePicker-single[_ngcontent-%COMP%]:after, .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.dateRangePicker-single[_ngcontent-%COMP%]:before, .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.mobile[_ngcontent-%COMP%]:after, .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.mobile[_ngcontent-%COMP%]:before{border-bottom-color:#fff}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]:after, .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]:before{border-bottom:10px solid #6e777c;border-left:10px solid transparent;border-right:10px solid transparent;content:\"\";height:0;position:absolute;top:-10px;width:0}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.open-left[_ngcontent-%COMP%]{left:0}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.open-left[_ngcontent-%COMP%]:before{left:10px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.open-left[_ngcontent-%COMP%]:after{display:none}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.open-right[_ngcontent-%COMP%]{right:0}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.open-right[_ngcontent-%COMP%]:before{display:none}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.open-right[_ngcontent-%COMP%]:after{right:10px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker.is-chrome[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{position:relative;top:1px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]   .btn[_ngcontent-%COMP%], .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]   .btn.btn-link[_ngcontent-%COMP%]{min-width:auto;width:auto}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]{background-color:#6e777c;border-top-left-radius:5px;border-top-right-radius:5px;color:#fff;font-size:16px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%] > .col[_ngcontent-%COMP%]{min-width:270px;padding:.5rem 15px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]{align-content:flex-start;align-items:center!important;display:flex}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]   i[_ngcontent-%COMP%]{color:#fff;margin-right:.5rem}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]   i.material-icons[_ngcontent-%COMP%]{top:0}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .input-with-icon[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{line-height:1}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .form-control[_ngcontent-%COMP%]{background:none;border:0;color:#fff;display:inline-block;padding:0;width:150px}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .form-control[_ngcontent-%COMP%]::-moz-placeholder{color:#fff}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]   .form-inputs[_ngcontent-%COMP%]   .form-control[_ngcontent-%COMP%]::placeholder{color:#fff}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]   .row-calendars[_ngcontent-%COMP%]   .col__mobile[_ngcontent-%COMP%]{border-top:1px dotted #ccc;margin-top:1rem}.dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]:not(.mobile)   .form-inputs[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]:nth-child(2), .dateRangePicker-wrapper[_ngcontent-%COMP%]   .dateRangePicker[_ngcontent-%COMP%]:not(.mobile)   .row-calendars[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]:nth-child(2){border-left:2px solid #c6d1d6}.dateRangePicker-wrapper[_ngcontent-%COMP%]   i.material-icons[_ngcontent-%COMP%]{position:relative;top:2px}"] });
DateRangePickerComponent.ctorParameters = () => [];
DateRangePickerComponent.propDecorators = {
    options: [{ type: Input }],
    controlName: [{ type: Input }],
    instanceId: [{ type: Input }],
    parentFormGroup: [{ type: Input }],
    fromDate: [{ type: Input }],
    toDate: [{ type: Input }],
    datepickerReset: [{ type: Output }],
    rangeSelected: [{ type: Output }],
    handleClick: [{ type: HostListener, args: ['document:click', ['$event'],] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DateRangePickerComponent, [{
        type: Component,
        args: [{
                encapsulation: ViewEncapsulation.Emulated,
                selector: 'date-range-picker',
                template: "<div class=\"dateRangePicker-wrapper\" *ngIf=\"parentFormGroup\">\n  <div #content>\n    <ng-content></ng-content>\n  </div>\n  <div class=\"form-group\" [formGroup]=\"parentFormGroup\">\n    <ng-container *ngIf=\"content.childNodes.length === 0\">\n      <label [attr.for]=\"instanceId\">{{ options.labelText }}</label>\n      <input [attr.id]=\"instanceId\" class=\"form-control\" type=\"text\" [formControlName]=\"controlName\">\n    </ng-container>\n\n    <div class=\"modal\" tabindex=\"-1\" role=\"dialog\" [ngClass]=\"{ 'modal-display': displayStyle}\">\n      <div class=\"modal-dialog modal-dialog-centered\" role=\"document\">\n        <div class=\"modal-content\">\n          <div class=\"modal-body\">\n            <div [ngClass]=\"{'dateRangePickerModal': true,  'dateRangePicker-single': options.singleCalendar, 'is-chrome': checkChrome(), 'mobile': isMobile }\">\n              <div class=\"container-fluid\">\n                <div class=\"row\">\n                  <div class=\"col\">\n        \n                    <!-- Manual Inputs -->\n                    <manual-inputs [options]=\"options\" [fromDate]=\"fromDate\" [toDate]=\"toDate\" [isMobile]=\"isMobile\" (dateFromInput)=\"setDateFromInput($event)\"></manual-inputs>\n        \n                    <!-- Calendars -->\n                    <div class=\"row row-calendars\">\n                      <div [ngClass]=\"{ 'col-12': true, 'col-sm-6': !isMobile && !options.singleCalendar }\">\n                        <calendar [ngClass]=\"{ 'calendar': true, 'calendar-single': options.singleCalendar }\" [isLeft]=\"true\" [month]=\"fromMonth\" [year]=\"fromYear\" (monthChanged)=monthChanged($event) (yearChanged)=\"yearChanged($event)\" (dateChanged)=\"dateChanged($event)\" [format]=\"options.format\" [selectedFromDate]=\"fromDate\" [selectedToDate]=\"toDate\" [minDate]=\"options.minDate\" [maxDate]=\"options.maxDate\" [singleCalendar]=\"options.singleCalendar\" [icons]=\"options.icons\"></calendar>\n                      </div>\n        \n                      <div [ngClass]=\"{ 'col-12': true, 'col-sm-6': !isMobile, 'col__mobile': isMobile }\" *ngIf=\"!options.singleCalendar\">\n                        <calendar class=\"calendar\" [month]=\"toMonth\" [year]=\"toYear\" [format]=\"options.format\" (dateChanged)=\"dateChanged($event)\" (monthChanged)=\"monthChanged($event)\" (yearChanged)=\"yearChanged($event)\" [selectedFromDate]=\"fromDate\" [selectedToDate]=\"toDate\" [minDate]=\"options.minDate\" [maxDate]=\"options.maxDate\" [icons]=\"options.icons\"></calendar>\n                      </div>\n                    </div>\n                  </div>\n                </div>\n        \n                <!-- Pre-Defined Date Ranges -->\n                <predefined-date-ranges [options]=\"options\" [range]=\"range\" [enableApplyButton]=\"enableApplyButton\" (closeEvent)=\"close($event)\" (resetEvent)=\"reset($event)\" (applyEvent)=\"apply($event)\" (predefinedRange)=\"applyPredefinedRange($event)\"></predefined-date-ranges>\n        \n              </div>\n            </div>\n          </div>\n          <div class=\"modal-footer\">\n            <button type=\"button\" class=\"btn btn-outline-secondary btn-sm\" (click)=\"close($event)\">Close</button>\n            <button type=\"button\" class=\"btn btn-outline-secondary btn-sm ml-3\" (click)=\"reset($event)\" *ngIf=\"range && options.showResetButton\">Reset</button>\n            <button type=\"button\" class=\"btn btn-primary btn-sm ml-3\" [disabled]=\"!enableApplyButton\" (click)=\"apply($event)\" *ngIf=\"!options.autoApply\">Apply</button>\n          </div>\n        </div>\n      </div>\n    </div>\n\n    <!-- Desktop Browsers -->\n    <div [ngClass]=\"{'dateRangePicker': true, 'invisible': !showCalendars, 'dateRangePicker-single': options.singleCalendar, 'is-chrome': checkChrome(), 'open-left': options.position === 'left', 'open-right': options.position === 'right', 'mobile': isMobile }\">\n      <div class=\"container-fluid\">\n        <div class=\"row\">\n          <div class=\"col\">\n\n            <!-- Manual Inputs -->\n            <manual-inputs [options]=\"options\" [fromDate]=\"fromDate\" [toDate]=\"toDate\" [isMobile]=\"isMobile\" (dateFromInput)=\"setDateFromInput($event)\"></manual-inputs>\n\n            <!-- Calendars -->\n            <div class=\"row row-calendars\">\n              <div [ngClass]=\"{ 'col-12': true, 'col-sm-6': !isMobile && !options.singleCalendar }\">\n                <calendar [ngClass]=\"{ 'calendar': true, 'calendar-single': options.singleCalendar }\" [isLeft]=\"true\" [month]=\"fromMonth\" [year]=\"fromYear\" (monthChanged)=monthChanged($event) (yearChanged)=\"yearChanged($event)\" (dateChanged)=\"dateChanged($event)\" [format]=\"options.format\" [selectedFromDate]=\"fromDate\" [selectedToDate]=\"toDate\" [minDate]=\"options.minDate\" [maxDate]=\"options.maxDate\" [singleCalendar]=\"options.singleCalendar\" [icons]=\"options.icons\"></calendar>\n              </div>\n\n              <div [ngClass]=\"{ 'col-12': true, 'col-sm-6': !isMobile, 'col__mobile': isMobile }\" *ngIf=\"!options.singleCalendar\">\n                <calendar class=\"calendar\" [month]=\"toMonth\" [year]=\"toYear\" [format]=\"options.format\" (dateChanged)=\"dateChanged($event)\" (monthChanged)=\"monthChanged($event)\" (yearChanged)=\"yearChanged($event)\" [selectedFromDate]=\"fromDate\" [selectedToDate]=\"toDate\" [minDate]=\"options.minDate\" [maxDate]=\"options.maxDate\" [icons]=\"options.icons\"></calendar>\n              </div>\n            </div>\n          </div>\n        </div>\n\n        <!-- Pre-Defined Date Ranges -->\n        <predefined-date-ranges [options]=\"options\" [range]=\"range\" [enableApplyButton]=\"enableApplyButton\" (closeEvent)=\"close($event)\" (resetEvent)=\"reset($event)\" (applyEvent)=\"apply($event)\"></predefined-date-ranges>\n\n      </div>\n    </div>\n\n  </div>\n</div>\n",
                styles: [".dateRangePicker-wrapper{border:none;position:relative}.dateRangePicker-wrapper input.form-control[disabled]{background-color:#fff}.dateRangePicker-wrapper .modal-body{padding:0!important}.dateRangePicker-wrapper .modal{-webkit-backdrop-filter:brightness(50%);backdrop-filter:brightness(50%)}.dateRangePicker-wrapper .modal-display{display:flex}.dateRangePicker-wrapper .modal-display .modal-dialog{box-shadow:none;display:flex}.dateRangePicker-wrapper .modal-content{border:0;border-radius:8px;box-shadow:0 0 10px 0 rgba(129,139,144,.5);display:flex;width:602px}.dateRangePicker-wrapper .modal-content .modal-footer .btn{width:auto}.dateRangePicker-wrapper .dateRangePickerModal{border-top-left-radius:5px;border-top-right-radius:5px;font-family:Roboto,Helvetica,Arial,sans-serif;overflow:visible;z-index:3000}@media screen and (min-width:576px){.dateRangePicker-wrapper .dateRangePickerModal{min-width:600px}}.dateRangePicker-wrapper .dateRangePickerModal.dateRangePicker-single,.dateRangePicker-wrapper .dateRangePickerModal.mobile{min-width:300px}.dateRangePicker-wrapper .dateRangePickerModal.dateRangePicker-single:after,.dateRangePicker-wrapper .dateRangePickerModal.dateRangePicker-single:before,.dateRangePicker-wrapper .dateRangePickerModal.mobile:after,.dateRangePicker-wrapper .dateRangePickerModal.mobile:before{border-bottom-color:#fff}.dateRangePicker-wrapper .dateRangePickerModal.is-chrome .form-inputs .input-with-icon span{position:relative;top:1px}.dateRangePicker-wrapper .dateRangePickerModal .btn,.dateRangePicker-wrapper .dateRangePickerModal .btn.btn-link{min-width:60px;width:60px}.dateRangePicker-wrapper .dateRangePickerModal .form-inputs{background-color:#6e777c;border-top-left-radius:5px;border-top-right-radius:5px;color:#fff;font-size:16px}.dateRangePicker-wrapper .dateRangePickerModal .form-inputs>.col{min-width:270px;padding:.5rem 15px}.dateRangePicker-wrapper .dateRangePickerModal .form-inputs .input-with-icon{align-content:flex-start;align-items:center!important;display:flex}.dateRangePicker-wrapper .dateRangePickerModal .form-inputs .input-with-icon i{color:#fff;margin-right:.5rem}.dateRangePicker-wrapper .dateRangePickerModal .form-inputs .input-with-icon i.material-icons{top:0}.dateRangePicker-wrapper .dateRangePickerModal .form-inputs .input-with-icon span{line-height:1}.dateRangePicker-wrapper .dateRangePickerModal .form-inputs .form-control{background:none;border:0;color:#fff;display:inline-block;padding:0;width:150px}.dateRangePicker-wrapper .dateRangePickerModal .form-inputs .form-control::-moz-placeholder{color:#fff}.dateRangePicker-wrapper .dateRangePickerModal .form-inputs .form-control::placeholder{color:#fff}.dateRangePicker-wrapper .dateRangePickerModal .row-1s .col__mobile{border-top:1px dotted #ccc;margin-top:1rem}.dateRangePicker-wrapper .dateRangePickerModal:not(.mobile) .form-inputs>div:nth-child(2),.dateRangePicker-wrapper .dateRangePickerModal:not(.mobile) .row-calendars>div:nth-child(2){border-left:2px solid #c6d1d6}.dateRangePicker-wrapper .dateRangePicker{background:#fff;border-radius:9px;box-shadow:0 0 10px 0 rgba(129,139,144,.5);font-family:Roboto,Helvetica,Arial,sans-serif;overflow:visible;position:absolute;top:88px;z-index:3000}@media screen and (min-width:576px){.dateRangePicker-wrapper .dateRangePicker{min-width:600px}}.dateRangePicker-wrapper .dateRangePicker.dateRangePicker-single,.dateRangePicker-wrapper .dateRangePicker.mobile{min-width:300px}.dateRangePicker-wrapper .dateRangePicker.dateRangePicker-single:after,.dateRangePicker-wrapper .dateRangePicker.dateRangePicker-single:before,.dateRangePicker-wrapper .dateRangePicker.mobile:after,.dateRangePicker-wrapper .dateRangePicker.mobile:before{border-bottom-color:#fff}.dateRangePicker-wrapper .dateRangePicker:after,.dateRangePicker-wrapper .dateRangePicker:before{border-bottom:10px solid #6e777c;border-left:10px solid transparent;border-right:10px solid transparent;content:\"\";height:0;position:absolute;top:-10px;width:0}.dateRangePicker-wrapper .dateRangePicker.open-left{left:0}.dateRangePicker-wrapper .dateRangePicker.open-left:before{left:10px}.dateRangePicker-wrapper .dateRangePicker.open-left:after{display:none}.dateRangePicker-wrapper .dateRangePicker.open-right{right:0}.dateRangePicker-wrapper .dateRangePicker.open-right:before{display:none}.dateRangePicker-wrapper .dateRangePicker.open-right:after{right:10px}.dateRangePicker-wrapper .dateRangePicker.is-chrome .form-inputs .input-with-icon span{position:relative;top:1px}.dateRangePicker-wrapper .dateRangePicker .btn,.dateRangePicker-wrapper .dateRangePicker .btn.btn-link{min-width:auto;width:auto}.dateRangePicker-wrapper .dateRangePicker .form-inputs{background-color:#6e777c;border-top-left-radius:5px;border-top-right-radius:5px;color:#fff;font-size:16px}.dateRangePicker-wrapper .dateRangePicker .form-inputs>.col{min-width:270px;padding:.5rem 15px}.dateRangePicker-wrapper .dateRangePicker .form-inputs .input-with-icon{align-content:flex-start;align-items:center!important;display:flex}.dateRangePicker-wrapper .dateRangePicker .form-inputs .input-with-icon i{color:#fff;margin-right:.5rem}.dateRangePicker-wrapper .dateRangePicker .form-inputs .input-with-icon i.material-icons{top:0}.dateRangePicker-wrapper .dateRangePicker .form-inputs .input-with-icon span{line-height:1}.dateRangePicker-wrapper .dateRangePicker .form-inputs .form-control{background:none;border:0;color:#fff;display:inline-block;padding:0;width:150px}.dateRangePicker-wrapper .dateRangePicker .form-inputs .form-control::-moz-placeholder{color:#fff}.dateRangePicker-wrapper .dateRangePicker .form-inputs .form-control::placeholder{color:#fff}.dateRangePicker-wrapper .dateRangePicker .row-calendars .col__mobile{border-top:1px dotted #ccc;margin-top:1rem}.dateRangePicker-wrapper .dateRangePicker:not(.mobile) .form-inputs>div:nth-child(2),.dateRangePicker-wrapper .dateRangePicker:not(.mobile) .row-calendars>div:nth-child(2){border-left:2px solid #c6d1d6}.dateRangePicker-wrapper i.material-icons{position:relative;top:2px}"]
            }]
    }], function () { return []; }, { options: [{
            type: Input
        }], controlName: [{
            type: Input
        }], instanceId: [{
            type: Input
        }], fromDate: [{
            type: Input
        }], toDate: [{
            type: Input
        }], datepickerReset: [{
            type: Output
        }], rangeSelected: [{
            type: Output
        }], handleClick: [{
            type: HostListener,
            args: ['document:click', ['$event']]
        }], parentFormGroup: [{
            type: Input
        }] }); })();
