// ngx-daterange/esm2015/ngx-daterange.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { CalendarComponent } from './components/calendar/calendar.component';
import { DateRangePickerComponent } from './components/datepicker/date-range-picker.component';
import { FormatMomentDatePipe } from './pipes/format-moment-date.pipe';
import { ManualInputsComponent } from './components/manual-inputs/manual-inputs.component';
import { PredefinedDateRangesComponent } from './components/predefined-date-ranges/predefined-date-ranges.component';
import * as ɵngcc0 from '@angular/core';
const declarations = [
    CalendarComponent,
    DateRangePickerComponent,
    FormatMomentDatePipe,
    ManualInputsComponent,
    PredefinedDateRangesComponent
];
export class NgxDateRangeModule {
}
NgxDateRangeModule.ɵfac = function NgxDateRangeModule_Factory(t) { return new (t || NgxDateRangeModule)(); };
NgxDateRangeModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgxDateRangeModule });
NgxDateRangeModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            ReactiveFormsModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxDateRangeModule, [{
        type: NgModule,
        args: [{
                declarations: [...declarations],
                exports: [...declarations],
                imports: [
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxDateRangeModule, { declarations: function () { return [CalendarComponent,
        DateRangePickerComponent,
        FormatMomentDatePipe,
        ManualInputsComponent,
        PredefinedDateRangesComponent]; }, imports: function () { return [CommonModule,
        FormsModule,
        ReactiveFormsModule]; }, exports: function () { return [CalendarComponent,
        DateRangePickerComponent,
        FormatMomentDatePipe,
        ManualInputsComponent,
        PredefinedDateRangesComponent]; } }); })();
