// ngx-plyr/esm5/lib/plyr.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { PlyrComponent } from './plyr/plyr.component';
import * as ɵngcc0 from '@angular/core';
var PlyrModule = /** @class */ (function () {
    function PlyrModule() {
    }
PlyrModule.ɵfac = function PlyrModule_Factory(t) { return new (t || PlyrModule)(); };
PlyrModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: PlyrModule });
PlyrModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PlyrModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    PlyrComponent,
                ],
                exports: [
                    PlyrComponent,
                ]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(PlyrModule, { declarations: function () { return [PlyrComponent]; }, exports: function () { return [PlyrComponent]; } }); })();
    return PlyrModule;
}());
export { PlyrModule };
