// ngx-daterange/esm2015/components/predefined-date-ranges/predefined-date-ranges.component.js
import { Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function PredefinedDateRangesComponent_div_2_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 6);
    ɵngcc0.ɵɵlistener("click", function PredefinedDateRangesComponent_div_2_button_2_Template_button_click_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r5); const range_r3 = restoredCtx.$implicit; const ctx_r4 = ɵngcc0.ɵɵnextContext(2); return ctx_r4.applyPredefinedRange($event, range_r3); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const range_r3 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(range_r3.name);
} }
function PredefinedDateRangesComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵelementStart(1, "div", 1);
    ɵngcc0.ɵɵtemplate(2, PredefinedDateRangesComponent_div_2_button_2_Template, 2, 1, "button", 5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.options.preDefinedRanges);
} }
function PredefinedDateRangesComponent_ng_container_3_button_3_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 10);
    ɵngcc0.ɵɵlistener("click", function PredefinedDateRangesComponent_ng_container_3_button_3_Template_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(2); return ctx_r8.reset($event); });
    ɵngcc0.ɵɵtext(1, "Reset");
    ɵngcc0.ɵɵelementEnd();
} }
function PredefinedDateRangesComponent_ng_container_3_button_4_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 11);
    ɵngcc0.ɵɵlistener("click", function PredefinedDateRangesComponent_ng_container_3_button_4_Template_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(2); return ctx_r10.apply($event); });
    ɵngcc0.ɵɵtext(1, "Apply");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("disabled", !ctx_r7.enableApplyButton);
} }
function PredefinedDateRangesComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "button", 7);
    ɵngcc0.ɵɵlistener("click", function PredefinedDateRangesComponent_ng_container_3_Template_button_click_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.close($event); });
    ɵngcc0.ɵɵtext(2, "Close");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, PredefinedDateRangesComponent_ng_container_3_button_3_Template, 2, 0, "button", 8);
    ɵngcc0.ɵɵtemplate(4, PredefinedDateRangesComponent_ng_container_3_button_4_Template, 2, 1, "button", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.range && ctx_r1.options.showResetButton);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r1.options.autoApply);
} }
export class PredefinedDateRangesComponent {
    constructor() {
        this.closeEvent = new EventEmitter();
        this.resetEvent = new EventEmitter();
        this.applyEvent = new EventEmitter();
        this.predefinedRange = new EventEmitter();
    }
    ngOnInit() {
    }
    applyPredefinedRange(event, definedDateRange) {
        this.predefinedRange.emit({ event: event, definedDateRange: definedDateRange });
    }
    close(event) {
        this.closeEvent.emit(event);
        event.stopPropagation();
    }
    reset(event) {
        this.resetEvent.emit(event);
        event.stopPropagation();
    }
    apply(event) {
        this.applyEvent.emit(event);
        event.stopPropagation();
    }
}
PredefinedDateRangesComponent.ɵfac = function PredefinedDateRangesComponent_Factory(t) { return new (t || PredefinedDateRangesComponent)(); };
PredefinedDateRangesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PredefinedDateRangesComponent, selectors: [["predefined-date-ranges"]], inputs: { options: "options", range: "range", enableApplyButton: "enableApplyButton" }, outputs: { closeEvent: "closeEvent", resetEvent: "resetEvent", applyEvent: "applyEvent", predefinedRange: "predefinedRange" }, decls: 4, vars: 2, consts: [[1, "row", "text-center", "my-3"], [1, "col"], ["class", "row mb-2", 4, "ngIf"], [4, "ngIf"], [1, "row", "mb-2"], ["type", "button", "class", "btn btn-link", 3, "click", 4, "ngFor", "ngForOf"], ["type", "button", 1, "btn", "btn-link", 3, "click"], ["type", "button", 1, "btn", "btn-outline-secondary", "btn-sm", 3, "click"], ["type", "button", "class", "btn btn-outline-secondary btn-sm ml-3", 3, "click", 4, "ngIf"], ["type", "button", "class", "btn btn-primary btn-sm ml-3", 3, "disabled", "click", 4, "ngIf"], ["type", "button", 1, "btn", "btn-outline-secondary", "btn-sm", "ml-3", 3, "click"], ["type", "button", 1, "btn", "btn-primary", "btn-sm", "ml-3", 3, "disabled", "click"]], template: function PredefinedDateRangesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵtemplate(2, PredefinedDateRangesComponent_div_2_Template, 3, 1, "div", 2);
        ɵngcc0.ɵɵtemplate(3, PredefinedDateRangesComponent_ng_container_3_Template, 5, 2, "ng-container", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.options.singleCalendar && ctx.options.preDefinedRanges && ctx.options.preDefinedRanges.length > 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.options.modal);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgForOf], styles: [".btn[_ngcontent-%COMP%], .btn.btn-link[_ngcontent-%COMP%]{min-width:60px}"] });
PredefinedDateRangesComponent.ctorParameters = () => [];
PredefinedDateRangesComponent.propDecorators = {
    options: [{ type: Input }],
    range: [{ type: Input }],
    enableApplyButton: [{ type: Input }],
    closeEvent: [{ type: Output }],
    resetEvent: [{ type: Output }],
    applyEvent: [{ type: Output }],
    predefinedRange: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PredefinedDateRangesComponent, [{
        type: Component,
        args: [{
                encapsulation: ViewEncapsulation.Emulated,
                selector: 'predefined-date-ranges',
                template: "<div class=\"row text-center my-3\">\n\t<div class=\"col\">\n\t\t<div class=\"row mb-2\" *ngIf=\"!options.singleCalendar && options.preDefinedRanges && options.preDefinedRanges.length > 0\">\n\t\t\t<div class=\"col\">\n\t\t\t\t<button type=\"button\" *ngFor=\"let range of options.preDefinedRanges\" class=\"btn btn-link\" (click)=\"applyPredefinedRange($event, range)\">{{ range.name }}</button>\n\t\t\t</div>\n\t\t</div>\n\t\t<ng-container *ngIf=\"!options.modal\">\n\t\t\t<button type=\"button\" class=\"btn btn-outline-secondary btn-sm\" (click)=\"close($event)\">Close</button>\n\t\t\t<button type=\"button\" class=\"btn btn-outline-secondary btn-sm ml-3\" (click)=\"reset($event)\" *ngIf=\"range && options.showResetButton\">Reset</button>\n\t\t\t<button type=\"button\" class=\"btn btn-primary btn-sm ml-3\" [disabled]=\"!enableApplyButton\" (click)=\"apply($event)\" *ngIf=\"!options.autoApply\">Apply</button>\n\t\t</ng-container>\n\t</div>\n</div>",
                styles: [".btn,.btn.btn-link{min-width:60px}"]
            }]
    }], function () { return []; }, { closeEvent: [{
            type: Output
        }], resetEvent: [{
            type: Output
        }], applyEvent: [{
            type: Output
        }], predefinedRange: [{
            type: Output
        }], options: [{
            type: Input
        }], range: [{
            type: Input
        }], enableApplyButton: [{
            type: Input
        }] }); })();
